display_list/geometry/dl_region.cc

```diff
@@ -60,189 +626,116 @@ std::vector<SkIRect> DlRegion::getRects(bool deband) const {
+}
 
-  for (const SkIRect& rect : rects) {
-    if (rect.isEmpty()) {
-      continue;
+bool DlRegion::spansIntersect(const Span* begin1,
+                              const Span* end1,
+                              const Span* begin2,
+                              const Span* end2) {
+  while (begin1 != end1 && begin2 != end2) {
+    if (begin1->right <= begin2->left) {
+      ++begin1;
+    } else if (begin2->right <= begin1->left) {
+      ++begin2;
+    } else {
+      return true;
     }
+  }
+  return false;
+}
 
-    int32_t y1 = rect.fTop;
-    int32_t y2 = rect.fBottom;
+bool DlRegion::intersects(const DlRegion& region) const {
+  if (isEmpty() || region.isEmpty()) {
+    return false;
+  }
 
-    for (size_t i = start_index; i < lines_.size() && y1 < y2; ++i) {
-      SpanLine& line = lines_[i];
+  auto our_complex = isComplex();
+  auto their_complex = region.isComplex();
+  auto bounds_intersect = SkIRect::Intersects(bounds_, region.bounds_);
 
-      if (rect.fTop >= line.bottom) {
-        start_index = i;
-        continue;
-      }
+  if (!our_complex && !their_complex) {
+    return bounds_intersect;
+  }
 
-      if (y2 <= line.top) {
-        insertLine(i, makeLine(y1, y2, rect.fLeft, rect.fRight));
-        mark_dirty(i);
-        y1 = y2;
-        break;
-      }
-      if (y1 < line.top) {
-        auto prevLineStart = line.top;
-        insertLine(i, makeLine(y1, prevLineStart, rect.fLeft, rect.fRight));
-        mark_dirty(i);
-        y1 = prevLineStart;
-        continue;
-      }
-      if (y1 > line.top) {
-        // duplicate line
-        auto prevLineEnd = line.bottom;
-        line.bottom = y1;
-        mark_dirty(i);
-        insertLine(i + 1, makeLine(y1, prevLineEnd, *line.spans));
-        continue;
-      }
-      FML_DCHECK(y1 == line.top);
-      if (y2 < line.bottom) {
-        // duplicate line
-        auto newLine = makeLine(y2, line.bottom, *line.spans);
-        line.bottom = y2;
-        line.insertSpan(rect.fLeft, rect.fRight);
-        insertLine(i + 1, newLine);
-        y1 = y2;
-        mark_dirty(i);
-        break;
-      }
-      FML_DCHECK(y2 >= line.bottom);
-      line.insertSpan(rect.fLeft, rect.fRight);
-      mark_dirty(i);
-      y1 = line.bottom;
-    }
+  if (!bounds_intersect) {
+    return false;
+  }
 
-    if (y1 < y2) {
-      lines_.push_back(makeLine(y1, y2, rect.fLeft, rect.fRight));
-      mark_dirty(lines_.size() - 1);
-    }
+  if (!our_complex) {
+    return region.intersects(bounds_);
+  }
 
-    // Check for duplicate lines and merge them.
-    if (dirty_start <= dirty_end) {
-      // Expand the region by one if possible.
-      if (dirty_start > 0) {
-        --dirty_start;
-      }
-      if (dirty_end + 1 < lines_.size()) {
-        ++dirty_end;
+  if (!their_complex) {
+    return intersects(region.bounds_);
+  }
+
+  auto ours = lines_.begin();
+  auto theirs = region.lines_.begin();
+
+  while (ours != lines_.end() && theirs != region.lines_.end()) {
```

Comment:
In the other loops you pulled the end() references out into local variables. Would that help here too?